modules/pam_setquota/pam_setquota.c

```diff
@@ -164,15 +165,12 @@ PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
   *mntdevice = '\0';
   while ((mnt = getmntent(fp)) != NULL) {
     if (param.fs == NULL) {
+      /* If fs is not specified use filesystem with homedir as default */
       size_t mnt_len = strlen(mnt->mnt_dir);
-      /* If fs is not specified use filesystem with homedir as default
-       * Checking the mnt_len-th character in pwd->pw_dir is safe because of the
-       * strncmp(2) check before (whose success implies strlen(pwd->pw_dir) >=
-       * mntlen)
-       */
-      if ((strncmp(pwd->pw_dir, mnt->mnt_dir, mnt_len) == 0) &&
-          (mnt_len > match_size) &&
-          (pwd->pw_dir[mnt_len] == '\0' || pwd->pw_dir[mnt_len] == '/')) {
+      const char *s;
+      if (mnt_len > match_size &&
+         (s = pam_str_skip_prefix_len(pwd->pw_dir, mnt->mnt_dir, mnt_len)) != NULL &&
+         (s[0] == '\0' || s[0] == '/')) {
```

Comment:
I need to go back on parts of that change, because, as it turns out, the code was incorrect from the start.
I need to look at `pwd->pw_dir[mnt_len-1]` to check for `\0` or `/` and this seems not possible with `pam_str_skip_prefix_len()`, because it would technically be `s[-1]`.

Reply:
Why? You need to check the symbol behind the matched prefix, don't you?

Reply:
No. I've been debugging this for half of the day.
It needs to find the longest matching mountpoint containing pwd->pw_dir.
The above code never triggers, even if /home/hartge is on /.

Reply:
Or maybe, and I just considered this now, *s needs to be initialized to NULL?

Reply:
> No. I've been debugging this for half of the day.
> It needs to find the longest matching mountpoint containing pwd->pw_dir.
> The above code never triggers, even if /home/hartge is on /.

Inspecting s for "/home/hartge and "/" and "1" as input results in s being "home/hartge".
But I need to check the last byte that matched and not the one after, so s contains the wrong stuff here.

Reply:
To add to my shame: That part of the code was broken for a long time and I never noticed it, because I always used it with explicit fs=... parameters and never fell back on the automatic home-directory search.

Reply:
You just have to decrease mnt_len for every trailing "/" in mnt->mnt_dir, that's all.

Reply:
Wouldn't that break in mnt_dir='/' and mnt_len then being 0?

Reply:
something like this:
```size_t mnt_len;
for (mnt_len = strlen(mnt->mnt_dir); mnt_len > 0; --mnt_len)
  if (mnt->mnt_dir[mnt_len - 1] != '/')
    break;```
This way ```mnt_len``` might become 0 so ```mnt_len > match_size``` condition has to be dealt with somehow.

Reply:
> > No. I've been debugging this for half of the day.
> > It needs to find the longest matching mountpoint containing pwd->pw_dir.
> > The above code never triggers, even if /home/hartge is on /.
> 
> Inspecting s for "/home/hartge and "/" and "1" as input results in s being "home/hartge".
> But I need to check the last byte that matched and not the one after, so s contains the wrong stuff here.

Please repeat the same thought experiment with "/home/hartge" and "/home".

Reply:
Yes, I am adding a bunch of logging code to the function to throw onto my test VM.
I just noticed there is something else strange here. I need to meditate over this for a night.

Reply:
From looking at mnt_dir, pw_dir and the resulting s I think I only need to special case mnt_dir='/', because it is the only mounted filesystem with a trailing /.
In every other case, the remaining s will always either start with a / or be empty, in case of a complete match.

Reply:
Yes.

Reply:
Ok, I got it working as it should now, the if() just needed the special case for mnt_len == 0 and that solved the problem of not finding pw_dir if it was contained in /.
Now it looks like this (I've snipped some uninteresting lines for brevity):
1) User "oweh" with $HOME in /home/oweh contained directly in /:
```
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/sys"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/proc"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/dev"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/dev/pts"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/run"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Found pw_dir="/home/oweh" in mnt_dir="/" with suffix="/home/oweh" on device="/dev/sda1"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/sys/kernel/security"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/dev/shm"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/run/lock"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/sys/fs/cgroup"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/sys/fs/cgroup/unified"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/home2"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/home3"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): Trying to match pw_dir="/home/oweh" with mnt_dir="/run/user/1000"
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): device=/dev/sda1 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
Apr 10 15:11:04 debian sshd[14343]: pam_setquota(sshd:session): device=/dev/sda1 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
```
2) User "tester" with $HOME in "/home2" on the fs "/home2":
```
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/sys"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/proc"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/dev"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/dev/pts"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/run"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Found pw_dir="/home2" in mnt_dir="/" with suffix="/home2" on device="/dev/sda1"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/sys/kernel/security"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/dev/shm"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/run/lock"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/sys/kernel/debug"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/run/user/117"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/run/user/0"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/home2"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Found pw_dir="/home2" in mnt_dir="/home2" with suffix="" on device="/dev/loop0"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/home3"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): Trying to match pw_dir="/home2" with mnt_dir="/run/user/1001"
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): device=/dev/loop0 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
Apr 10 15:11:43 debian sshd[14377]: pam_setquota(sshd:session): device=/dev/loop0 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
```
3) And finally user "test3" with $HOME as "/home3/test3" in fs "/home3":
```
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/sys"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/proc"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/dev"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/dev/pts"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Found pw_dir="/home3/test3" in mnt_dir="/" with suffix="/home3/test3" on device="/dev/sda1"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/sys/kernel/security"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/dev/shm"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run/lock"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/sys/fs/cgroup"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/sys/kernel/debug"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run/user/117"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run/user/0"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/home2"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/home3"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Found pw_dir="/home3/test3" in mnt_dir="/home3" with suffix="/test3" on device="/dev/loop1"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run/user/1001"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): Trying to match pw_dir="/home3/test3" with mnt_dir="/run/user/1002"
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): device=/dev/loop1 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
Apr 10 15:12:18 debian sshd[14407]: pam_setquota(sshd:session): device=/dev/loop1 bsoftlimit=512000 bhardlimit=563200 isoftlimit=0 ihardlimit=0
```